compiler/next/lib/resolution/resolution-queries.cpp

```diff
@@ -33,313 +35,1441 @@
+      scopeStack.push_back(scopeForId(context, ast->id()));
+    }
+  }
+  void exitScope(const ASTNode* ast) {
+    if (createsScope(ast->tag())) {
+      scopeStack.pop_back();
+    }
+  }
+
+  bool enter(const Literal* literal) {
+    ResolvedExpression& result = byPostorder.byAst(literal);
+    result.type = typeForLiteral(context, literal);
+    return false;
+  }
+  void exit(const Literal* literal) {
+  }
+
+  bool enter(const Identifier* ident) {
+    assert(scopeStack.size() > 0);
+    const Scope* scope = scopeStack.back();
+    ResolvedExpression& result = byPostorder.byAst(ident);
+
+    LookupConfig config = LOOKUP_DECLS |
+                          LOOKUP_IMPORT_AND_USE |
+                          LOOKUP_PARENTS |
+                          LOOKUP_INNERMOST;
+
+    auto vec = lookupInScope(context, scope, ident, config);
+    if (vec.size() == 0) {
+      result.type = QualifiedType(QualifiedType::UNKNOWN, nullptr);
+    } else if (vec.size() > 1) {
+      // can't establish the type. If this is in a function
+      // call, we'll establish it later anyway.
     } else {
-      keep.expr = addin.expr;
-      keep.decl = addin.decl;
-      keep.type = addin.type;
-      keep.otherFns.swap(addin.otherFns);
-      return true; // updated
+      // vec.size() == 1
+      ID id = vec[0].firstId();
+      QualifiedType type;
+      if (id.isEmpty()) {
+        // empty IDs from the scope resolution process are builtins
+        type = typeForBuiltin(context, ident->name());
+      } else {
+        // use the type established at declaration/initialization
+        type = typeForId(id);
+      }
+      result.toId = id;
+      result.type = type;
+      // if there are multiple ids we should have gotten
+      // a multiple definition error at the declarations.
     }
+    return false;
+  }
+  void exit(const Identifier* ident) {
   }
-};
 
-template<> struct update<resolution::ResolvedSymbol> {
-  bool operator()(resolution::ResolvedSymbol& keep,
-                  resolution::ResolvedSymbol& addin) const {
-    bool match = keep.decl == addin.decl &&
-                 keep.typeSubs == addin.typeSubs &&
-                 keep.paramSubs == addin.paramSubs &&
-                 keep.instantiationPoint == addin.instantiationPoint &&
-                 keep.resolutionById.size() == addin.resolutionById.size();
-
-    if (match) {
-      // check also the resolutionById - we know sizes match here.
-      size_t n = keep.resolutionById.size();
-      for (size_t i = 0; i < n; i++) {
-        resolution::ResolutionResult& keepR = keep.resolutionById[i];
-        resolution::ResolutionResult& addinR = addin.resolutionById[i];
-        if (keepR.expr == addinR.expr &&
-            keepR.decl == addinR.decl &&
-            keepR.type == addinR.type &&
-            keepR.otherFns == addinR.otherFns) {
-          // OK, it matches
-        } else {
-          match = false;
-          break;
+  bool enter(const NamedDecl* decl) {
```

Comment:
We'll see it on the second definition for the same symbol. (And again if there's a third or fourth etc). We won't see it on subsequent uses because the variable resolver process just picks the first match. I think this is reasonable but we haven't really gotten to the point of expecting good error messages in this resolver yet. So it's just a start.